Add Compatibility With Woocommerce 6.6

lovo-h opened the pull request:
<!--
Did I add a title? A descriptive, yet concise, title.
-->

<!--
Issue: Link to the GitHub issue this PR addresses (if appropriate).
-->

Fixes n/a.

## Changes proposed in this Pull Request:

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for the PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR allows WCStripe's form field settings to be compatible with WooCommerce 6.6. 

## Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Please follow the following guidelines when writing testing instructions:

- Include screenshots if there is no similar flow in the critical flows: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Critical-flows
- Assume instructions will be copied over to the Release Testing Instructions wiki page: https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions
- Assume instructions will be followed by external testers.
- Assume tester does not have intimate knowledge of Stripe.
-->

0. Ensure WooCommerce 6.6 has been installed.
1. As a merchant, navigate to `WooCommerce > Settings > Payments` and click on the `Stripe` row.
2. Click on the `Settings` tab.
3. Update the `Name` and `Description` fields and click `Save changes`.

![image](https://user-images.githubusercontent.com/6844516/173426011-d47ab9f7-969e-41cd-b531-df4ce87f70e1.png)

5. As a customer, navigate to the shop, add an item to your cart, and navigate to checkout.
6. Ensure the changes to the `Name` and `Description` fields are reflected in the payment section.

![image](https://user-images.githubusercontent.com/6844516/173425918-4e1d0779-df03-4bd7-a611-5210d21550c4.png)

---

-   [x] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [x] Tested on mobile (or does not apply)

**Post merge**

-   [x] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)

lovo-h committed:
Fix minor comment typo: 'Processses' to 'Processes'

lovo-h committed:
Add 'short_statement_descriptor' form field setting

lovo-h committed:
Update 'title' & 'title_upe' fields' type

lovo-h committed:
Make form field settings compatible with WC v6.6

lovo-h committed:
Update changelog: 'Fix - Add compatibility with Woocommerce 6.6.'

asumaran commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -276,11 +276,13 @@ public function get_settings() {
 	 * @param WP_REST_Request $request Full data about the request.
 	 */
 	public function update_settings( WP_REST_Request $request ) {
+		$form_fields = $this->gateway->get_form_fields();
```
Wondering if it was necessary to get the `$form_fields` here. It seems possible to move it to the `validate_field()` method.  It's the actual place that it'll be used and we could avoid passing it to all methods below.

lovo-h replied:
I started off by placing `$this->gateway->get_form_fields()` in the `validate_field()` function but ultimately decided against it because I figured it would add a bit of overhead. Namely, each time `$this->gateway->get_form_fields()` is called, the following happens:
* Maps each element in `$form_fields` to ensure the `default` key is set.
* Calls `apply_filters( 'woocommerce_settings_api_form_fields_stripe', $form_fields )`.

If we did this for each form field, it would result in an `O(n^2)` runtime. So, I figured if we called it once, from `update_settings( ... )`, it would avoid this overhead (`O(n)` runtime). But since we do not need to call `get_form_fields()` for each form field, it should be ok to place it in `validate_field()` and tradeoff a bit of performance for code readability.

lovo-h replied:
Updated in [ee8bc59](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/commits/ee8bc5928feae59997a8f51f25b9568b766e909b).

asumaran commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -328,45 +330,79 @@ private function update_is_stripe_enabled( WP_REST_Request $request ) {
+	 *
+	 * @return string validated field value.
+	 */
+	private function validate_field( $field_key, $field_value, $form_fields ) {
+		if ( is_callable( [ $this->gateway, 'validate_' . $field_key . '_field' ] ) ) {
+			return $this->gateway->{'validate_' . $field_key . '_field'}( $field_key, $field_value );
+		}
+
+		if ( key_exists( $field_key, $form_fields ) ) {
+			$field_type = $form_fields[ $field_key ]['type'];
+
+			if ( is_callable( [ $this->gateway, 'validate_' . $field_type . '_field' ] ) ) {
+				return $this->gateway->{'validate_' . $field_type . '_field'}( $field_key, $field_value );
+			}
+		}
+
+		return $this->gateway->validate_text_field( $field_key, $field_value );
```
💯

asumaran commented on includes/admin/stripe-settings.php:
```diff
@@ -15,14 +15,14 @@
 		'title'                               => [
 			'title'       => __( 'Title', 'woocommerce-gateway-stripe' ),
-			'type'        => 'text',
+			'type'        => 'safe_text',
```
Does `safe_text` makes a difference with `text`? I've tested both and haven't found any.

lovo-h replied:
Yes, it does. In WooCommerce v6.6, `safe_text` has a separate validation function than `text`. The validation function call can be seen in [class-wc-rest-stripe-settings-controller.php::validate_field( ... )](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/commits/fb7c7b0374494176c4b862af8a51823c0df2c937#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR365):
```php
return $this->gateway->{'validate_' . $field_type . '_field'}( $field_key, $field_value );
```

reykjalin replied:
So what happens in WC < 6.6? Is `safe_text` even recognized? Do we need to make this a `WC_VERSION < 6.6 ? 'text' : 'safe_text'` type thing? 🤔

asumaran replied:
I seems it'll default to `text` https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR366

asumaran replied:
What I understand is that if the user upgrades WCStripe the issue will still exist until the form is saved again. But if the user stays with their WCStripe version and update WC to 6.6+ the fix will be applied at the moment the title is displayed.

It would be nice to fix for the case the user upgrades WCStripe only and doesn't save/update the form yet.

reykjalin replied:
> I seems it'll default to text https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/files#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR366

I thought this was called regardless of what the type of the field is? I think the `safe_text` is only detected by WC >= 6.6 where those versions of WC will apply sanitation _in addition to_ whatever validation/sanitation we do. I doubt they've updated older versions of WC to handle the `safe_text` type.

Have we confirmed that WC < 6.6 doesn't just skip the WC core validation/sanitation entirely if we use `safe_text`?

lovo-h replied:
> So what happens in WC < 6.6? Is `safe_text` even recognized? Do we need to make this a `WC_VERSION < 6.6 ? 'text' : 'safe_text'` type thing? 🤔

That's correct, `safe_text` is not recognized in WC < 6.6. But that's already been handled in this PR, like follows:
```php
private function validate_field( $field_key, $field_value, $form_fields ) {
  ...
  if ( key_exists( $field_key, $form_fields ) ) {
    $field_type = $form_fields[ $field_key ]['type'];
    // Checks if `validate_safe_text_field` exists.
    if ( is_callable( [ $this->gateway, 'validate_' . $field_type . '_field' ] ) ) {
      return $this->gateway->{'validate_' . $field_type . '_field'}( $field_key, $field_value );
    }
  }
  // Otherwise, defaults to `validate_text_field`.
  return $this->gateway->validate_text_field( $field_key, $field_value );
}
```
In other words, we do not need to apply the suggested ternary since `validate_field( ... )` defaults to `validate_text_field( ... )` already. [See code-block for more details](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/commits/fb7c7b0374494176c4b862af8a51823c0df2c937#diff-679186f0c3e522393b5394f6e65890f1b2bff5ef64efb250eee90d6348dc841eR356-R370).

> I thought this was called regardless of what the type of the field is? I think the safe_text is only detected by WC >= 6.6 where those versions of WC will apply sanitation in addition to whatever validation/sanitation we do. I doubt they've updated older versions of WC to handle the safe_text type.

Correct, they have not updated older versions of WC to include `safe_text` but as stated above, in that event, this PR defaults to `validate_text_field( ... )`. In other words, it treats it as if the type was `text`.

> Have we confirmed that WC < 6.6 doesn't just skip the WC core validation/sanitation entirely if we use safe_text?

Yes, we can confirm that if we use `safe_text` as the type, it will not skip validation. This can be seen in `validate_field( ... )`: it will default to `validate_text_field( ... )`. 

Edit: To elaborate a bit on this, the `stripe-settings.php` file is used by two other files:
1. `class-wc-gateway-stripe.php`
2. `class-wc-stripe-upe-payment-gateway.php`

These two files (payment gateways) retrieve and set the `title` (`title_upe`) by doing a `$this->get_option( ... )` call. Then `wordpress/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-payment-gateway.php` retrieves the title like follows:
```php
public function get_title() {
  return apply_filters( 'woocommerce_gateway_title', $this->title, $this->id );
}
```
But the latest version of WC retrieves the title like follows:
```php
public function get_title() {
  $title = wc_get_container()->get( HtmlSanitizer::class )->sanitize( $this->title, HtmlSanitizer::LOW_HTML_BALANCED_TAGS_NO_LINKS );
  return apply_filters( 'woocommerce_gateway_title', $title, $this->id );
}
```
All this to say that the form field's type was not considered in the flow.

Although, in `wordpress/wp-content/plugins/woocommerce/includes/abstracts/abstract-wc-settings-api.php::generate_settings_html( ... )` does consider the form field's type, it defaults to `text`.

lovo-h replied:
> What I understand is that if the user upgrades WCStripe the issue will still exist until the form is saved again. But if the user stays with their WCStripe version and update WC to 6.6+ the fix will be applied at the moment the title is displayed.
> 
> It would be nice to fix for the case the user upgrades WCStripe only and doesn't save/update the form yet.

Good call. We could do something similar to what WC is doing and apply the validation at the source of the output. In other words, we could validate at [class-wc-rest-stripe-settings-controller.php::get_settings()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/class-wc-rest-stripe-settings-controller.php#L237-L271). Thoughts?

lovo-h replied:
@reykjalin, although I do not think we need it, it wouldn't hurt to add the `WC_VERSION < 6.6 ? 'text' : 'safe_text'`. Added in [a752553](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/commits/a752553edd0d3b24cd959c0a765375606d07e731).

reykjalin replied:
> We could do something similar to what WC is doing and apply the validation at the source of the output. In other words, we could validate at [class-wc-rest-stripe-settings-controller.php::get_settings()](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/includes/admin/class-wc-rest-stripe-settings-controller.php#L237-L271). Thoughts?

I think applying the sanitation at the source of the output is a good idea and seems to be a general WP development recommendation. That is, of course, unless we're using this all over the place 😛 . In that case it'd probably be best to try to sanitize just once ☺️

asumaran reviewed (commented):
@lovo-h changes look good and work as expected. I'm leaving a couple of comments for you.

I wanted to mention that this PR doesn't fix the issue retroactively. Saved titles will still contain the problem until the Stripe settings form it's saved again.

Is it worth working on fixing the issue retroactively? Maybe we could update the saved form field at the moment the plugin is updated. If that's the case, @reykjalin worked on something similar on this [PR](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1672/files#diff-e1e5e9b442c2b5d4105edbdc3787651333c1923dd623677025f72d25796900d6R232)

The other option is to escape the unwanted HTML where the relevant inputs are used/displayed. Personally, I prefer the previous option.

lovo-h committed:
Fix typo: 'Woocommerce' to 'WooCommerce' in changelog

lovo-h committed:
Move $form_fields from 'update_settings()' to 'validate_field()'

Improves code readability but trades-off a bit of performance.

lovo-h commented:
> I wanted to mention that this PR doesn't fix the issue retroactively. Saved titles will still contain the problem until the Stripe settings form it's saved again.

@asumaran, I ran into this as well and forgot to bring it up in this PR. This is actually a separate issue that’s due to the React state. If you save the Stripe settings once and refresh the page, you will see that the changes were in fact saved the first time. We may need to create a separate issue for the React state not updating after the backend has updated.

Edit: Ah, upon re-reading your comment, it seems like you meant that the data already in the DB won't be affected by this PR. It seems like WooCommerce 6.6 has addressed that by validating the data before displaying it.

asumaran commented on includes/admin/class-wc-rest-stripe-settings-controller.php:
```diff
@@ -336,7 +336,34 @@ private function update_title( WP_REST_Request $request ) {
+			return $this->gateway->{'validate_' . $field_key . '_field'}( $field_key, $field_value );
+		}
+
+		$form_fields = $this->gateway->get_form_fields();
```
I don't see a problem with this but if we want to improve performance even further we could do the following:

```php
if ( ! $this->form_fields ) {
  $this->form_fields = $this->gateway->get_form_fields(); 
}
```

But the way it's done now works for me.

asumaran reviewed (approved):
🚀 Code looks good and tests as expected. I'm leaving just a non-blocking comment. Feel free to address it or leave it.

lovo-h committed:
Ensure to only use 'safe_text' for WC version 6.6 and greater

reykjalin reviewed (approved):
LGTM! All my reservations have been thoroughly addressed 🚀

asumaran commented:
@lovo-h @reykjalin The issue remains with WC 6.5.1 (until the user saves the Stripe settings form again). Is it worth fixing for older versions of WC?

lovo-h committed:
Retroactively apply validation on plugin settings

Before, validation would be applied only on settings that would be saved
after this commit has been merged. Now, validation will be applied on
settings regardless of whether they're saved, after this commit has been
merged.

lovo-h commented:
> @lovo-h @reykjalin The issue remains with WC 6.5.1 (until the user saves the Stripe settings form again). Is it worth fixing for older versions of WC?

Definitely worth addressing. I applied the solution mentioned in [this comment](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362#discussion_r897997193). Changes can be seen on [b0721be](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2362/commits/b0721be2f2a120399c236104dbff3b4d9a6d99c6).

asumaran reviewed (approved):
🚀 LGTM! Tests well with WC 6.5.1 and WC 6.6.0.
